File: leetcode/algorithms/606_construct_string_from_binary_tree/main.py
from typing import Optional


class Solution:
    class TreeNode:
        def __init__(self, val=0, left=None, right=None):
            self.val = val
            self.left = left
            self.right = right

    def tree2str(self, root: Optional[TreeNode]) -> str:
        if root == None:
            return ""

        if root.left == None and root.right == None:
            return f"{root.val}"

        left = "()"
        if root.left != None:
            left = f"({self.tree2str(root.left)})"

        right = ""
        if root.right != None:
            right = f"({self.tree2str(root.right)})"

        return f"{root.val}{left}{right}"

    # Best Solution
    def bestSolution(self, root: Optional[TreeNode]) -> str:
        return ""

File: leetcode/algorithms/606_construct_string_from_binary_tree/main.dart
class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}

class ConstructStringFromBinaryTree {
  String tree2str(TreeNode? root) {
    if (root == null) {
      return "";
    }

    String result = root.val.toString();

    if (root.left != null) {
      result += "(${tree2str(root.left)})";
    }

    if (root.right != null) {
      if (root.left == null) {
        result += "()";
      }
      result += "(${tree2str(root.right)})";
    }

    return result;
  }

  // Solution 1: DFS - Without recursion
  String solution1(TreeNode? root) {
    if (root == null) {
      return '';
    }

    // Create an empty stack
    final stack = <TreeNode>[];
    stack.add(root);

    // Create a set to keep track of visited nodes
    final visited = <TreeNode>{};

    final result = StringBuffer();

    while (stack.isNotEmpty) {
      final node = stack.last;

      if (visited.contains(node)) {
        stack.removeLast();
        result.write(')');
      } else {
        visited.add(node);
        result.write('(${node.val}');

        if (node.left == null && node.right != null) {
          result.write('()');
        }

        if (node.right != null) {
          stack.add(node.right!);
        }

        if (node.left != null) {
          stack.add(node.left!);
        }
      }
    }

    // Remove the leading '(' and trailing ')'
    return result.toString().substring(1, result.length - 1);
  }

  // Solution 2
  String solution2(TreeNode? root) {
    if (root == null) {
      return '';
    }

    String result = '${root.val}';

    if (root.left != null || root.right != null) {
      result += '(${tree2str(root.left)})';
    }

    if (root.right != null) {
      result += '(${tree2str(root.right)})';
    }

    return result;
  }
}
